pkg/services/ngalert/state/state.go

```diff
@@ -400,19 +395,38 @@ func resultKeepLast(state *State, rule *models.AlertRule, result eval.Result, lo
 	}
 }
 
-func (a *State) NeedsSending(resendDelay time.Duration) bool {
-	switch a.State {
-	case eval.Pending:
-		// We do not send notifications for pending states
+// NeedsSending returns true if the given state needs to be sent to the Alertmanager.
+// Reasons for sending include:
+// - The state has been resolved since the last notification.
+// - The state is firing and the last notification was sent at least resendDelay ago.
+// - The state was resolved within the resolvedRetention period, and the last notification was sent at least resendDelay ago.
+func (a *State) NeedsSending(resendDelay time.Duration, resolvedRetention time.Duration) bool {
+	if a.State == eval.Pending {
+		// We do not send notifications for pending states.
+		return false
+	}
+
+	// We should send a notification if the state has been resolved since the last notification.
+	if a.ResolvedAt.After(a.LastSentAt) {
+		return true
+	}
+
+	// For normal states, we should only be sending if this is a resolved notification or a re-send of the resolved
+	// notification within the resolvedRetention period.
+	if a.State == eval.Normal && !a.RecentlyResolved(resolvedRetention) {
 		return false
-	case eval.Normal:
-		// We should send a notification if the state is Normal because it was resolved
-		return a.Resolved
-	default:
-		// We should send, and re-send notifications, each time LastSentAt is <= LastEvaluationTime + resendDelay
-		nextSent := a.LastSentAt.Add(resendDelay)
-		return nextSent.Before(a.LastEvaluationTime) || nextSent.Equal(a.LastEvaluationTime)
 	}
+
+	// We should send, and re-send notifications, each time LastSentAt is <= LastEvaluationTime + resendDelay.
+	// This can include normal->normal transitions that were resolved in recent past evaluations.
+	nextSent := a.LastSentAt.Add(resendDelay)
+	return nextSent.Before(a.LastEvaluationTime) || nextSent.Equal(a.LastEvaluationTime)
+}
+
+// RecentlyResolved returns true if the state has been resolved within the resolvedRetention period. Will return false
+// if the state is not resolved or if the state was resolved outside the resolvedRetention period.
+func (a *State) RecentlyResolved(resolvedRetention time.Duration) bool {
+	return !a.ResolvedAt.IsZero() && a.LastEvaluationTime.Sub(a.ResolvedAt) <= resolvedRetention
```

Comment:
I think this can be a problem if rule evaluation interval is greater than resolvedRetention period. 
On other hand, I am not sure that there is a benefit of resending after such a long interval.

Reply:
This is covered by check in NeedsResending. WDYT about moving it here just to keep all logic related to resending resolved alert in one place?

Reply:
Sure, I can inline it. It was easier to grok while developing like this, but now that it's fully commented it feels unnecessary.